subsys/net/l2/ppp/fsm.c

```diff
@@ -509,14 +507,8 @@ int ppp_send_pkt(struct ppp_fsm *fsm, struct net_if *iface,
 		 * have returned from this function. That is bad because the
 		 * fsm would be in wrong state and the received pkt is dropped.
 		 */
-		fsm->sender.pkt = pkt;
-
-		/* FIXME: qemu_x86 crashes if timeout is 0 when running ppp
-		 * driver unit test. As a workaround set the timeout to 1 msec
-		 * in that case.
-		 */
-		(void)k_delayed_work_submit(&fsm->sender.work,
-			  IS_ENABLED(CONFIG_NET_TEST) ? K_MSEC(1) : K_NO_WAIT);
+		k_work_init(net_pkt_work(pkt), ppp_pkt_send);
```

Comment:
The work in net_pkt is used when the net_pkt is submitted to traffic queue to be sent out. Re-using the work for PPP looks clever, just wondering how well you have tested this as there might be some unseen interactions in the system if we reuse the variable for different purposes.

Reply:
According to `k_work_submit()` documentation work can be resubmitted in the handler. Implementation of `z_work_q_main()` also indicates that once handler is executed, there are no remaining information about the previous workqueue. So we use those facts and in the `ppp_pkt_send()` (which is work handler in PPP stack) we resubmit packet's work, but on the other workqueue using `net_if_queue_tx()`.

Please note that we are just modifying `ppp_send_pkt()` implementation to use net_pkt's work. This function is used only by link/network layer requests during negotiation (Conf-Req, Conf-Ack, etc.), later this packet is queued on interface (which is PPP), then to PPP serial driver and packed inside HDLC. TCP packets go to interface directly, so `ppp_send_pkt()` is not used there.

Regarding testing... this specific change works for few months on devices with modem, I had just not enough time in the meantime to rebase and publish change on GH. So I am pretty confident about it. From theoretical point of view I see no issues as well. In short we allocate the packet in PPP stack, so there are no other users yet. Then we submit packet to network stack (traffic queue mentioned by you) and don't use it ourselves anymore.